[inductor] Fix needs_fixed_stride_order silent incorrectness

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133639

Fixes #128084

The approach is option 2 of what Elias suggested in the comment
thread:
- We require tensors to have the correct stride at usage. This may
  involve a clone; if there was a clone and then a mutation into it
  then we copy_ back the result of the mutation.

The reason why I went this approach was because it was the easiest and
Inductor already works really hard to remove additional clones/copy_.

There are some cases that this doesn't generate efficient code for; for
example, if the tensor is a view, we don't change the base of the view
to have the right stride order, instead we do a clone.
The view case isn't very common so I'm ignoring it for now but we could
improve this in the future.

Test Plan:
- tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

zou3519 committed:
[inductor] Fix needs_fixed_stride_order silent incorrectness

Fixes #128084

The approach is two-fold (based on what Elias suggested in the comment
thread):
1. We realize Tensors and ensure they are the right stride.
2. If we can't ensure they have the right stride (if the tensors are
   already in an incorrect stride with FixedLayout, or if there's a
   view), then we require the correct stride on use via a clone and then
   copy_ back the result of the mutation.

We can improve this based on use cases (if the tensor is a view, then we
can require the base of the view to have the right stride order). The
view case isn't very common so I'm ignoring it for now.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] Fix needs_fixed_stride_order silent incorrectness"


Fixes #128084

The approach is two-fold (based on what Elias suggested in the comment
thread):
1. We realize Tensors and ensure they are the right stride.
2. If we can't ensure they have the right stride (if the tensors are
   already in an incorrect stride with FixedLayout, or if there's a
   view), then we require the correct stride on use via a clone and then
   copy_ back the result of the mutation.

We can improve this based on use cases (if the tensor is a view, then we
can require the base of the view to have the right stride order). The
view case isn't very common so I'm ignoring it for now.

Also, inductor is really good at removing unnecessary copy_: I couldn't
find a case where it the output code had it.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] Fix needs_fixed_stride_order silent incorrectness"


Fixes #128084

The approach is option 2 of what Elias suggested in the comment
thread:
- We require tensors to have the correct stride at usage. This may
  involve a clone; if there was a clone and then a mutation into it
  then we copy_ back the result of the mutation.

The reason why I went this approach was because it was the easiest and
Inductor already works really hard to remove additional clones/copy_.

There are some cases that this doesn't generate efficient code for; for
example, if the tensor is a view, we don't change the base of the view
to have the right stride order, instead we do a clone.
The view case isn't very common so I'm ignoring it for now but we could
improve this in the future.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] Fix needs_fixed_stride_order silent incorrectness"


Fixes #128084

The approach is option 2 of what Elias suggested in the comment
thread:
- We require tensors to have the correct stride at usage. This may
  involve a clone; if there was a clone and then a mutation into it
  then we copy_ back the result of the mutation.

The reason why I went this approach was because it was the easiest and
Inductor already works really hard to remove additional clones/copy_.

There are some cases that this doesn't generate efficient code for; for
example, if the tensor is a view, we don't change the base of the view
to have the right stride order, instead we do a clone.
The view case isn't very common so I'm ignoring it for now but we could
improve this in the future.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] Fix needs_fixed_stride_order silent incorrectness"


Fixes #128084

The approach is option 2 of what Elias suggested in the comment
thread:
- We require tensors to have the correct stride at usage. This may
  involve a clone; if there was a clone and then a mutation into it
  then we copy_ back the result of the mutation.

The reason why I went this approach was because it was the easiest and
Inductor already works really hard to remove additional clones/copy_.

There are some cases that this doesn't generate efficient code for; for
example, if the tensor is a view, we don't change the base of the view
to have the right stride order, instead we do a clone.
The view case isn't very common so I'm ignoring it for now but we could
improve this in the future.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] Fix needs_fixed_stride_order silent incorrectness"


Fixes #128084

The approach is option 2 of what Elias suggested in the comment
thread:
- We require tensors to have the correct stride at usage. This may
  involve a clone; if there was a clone and then a mutation into it
  then we copy_ back the result of the mutation.

The reason why I went this approach was because it was the easiest and
Inductor already works really hard to remove additional clones/copy_.

There are some cases that this doesn't generate efficient code for; for
example, if the tensor is a view, we don't change the base of the view
to have the right stride order, instead we do a clone.
The view case isn't very common so I'm ignoring it for now but we could
improve this in the future.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] Fix needs_fixed_stride_order silent incorrectness"


Fixes #128084

The approach is option 2 of what Elias suggested in the comment
thread:
- We require tensors to have the correct stride at usage. This may
  involve a clone; if there was a clone and then a mutation into it
  then we copy_ back the result of the mutation.

The reason why I went this approach was because it was the easiest and
Inductor already works really hard to remove additional clones/copy_.

There are some cases that this doesn't generate efficient code for; for
example, if the tensor is a view, we don't change the base of the view
to have the right stride order, instead we do a clone.
The view case isn't very common so I'm ignoring it for now but we could
improve this in the future.

Test Plan:
- tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on torch/_inductor/graph.py:
```diff
@@ -1236,6 +1213,53 @@ def significant_strides_equal(
+        # Those do not have lowerings pre-registered.
+        if op in lowerings:
+            return
+        # lowering an op that is FallbackKernel for the first time will automatically
+        # generate the lowering so that the following isn't called multiple times.
+        if torch._C.Tag.needs_fixed_stride_order in op.tags:
+            torch._inductor.lowering.add_layout_constraint(op, constrain_to_fx_strides)
```
where is the logic that prevents this from being called twice ? tried to find and failed

zou3519 replied:
I'll make the comment clearer. The first time we call maybe_add_custom_op_layout_constraints with a custom op `foo`:
- there's no lowering for the custom op
- directly after this function, in run_node, we do GraphLowering.call_function
- This generates a lowering for the custom op (make_fallback in https://github.com/pytorch/pytorch/blob/2bffbe06bdf8f130159917c3ecda2c5852fd35b4/torch/_inductor/graph.py#L1023)

The second time we call maybe_add_custom_op_layout_constraints with a custom op `foo`: 
- there's a lowering for the custom op, so the early return prevents us from adding the layout constraint twice.

Note that this was how the code worked previously [(we would generate the lowering and add the layout constraint once on the first usage of the custom op)](https://github.com/pytorch/pytorch/blob/2bffbe06bdf8f130159917c3ecda2c5852fd35b4/torch/_inductor/graph.py#L1010-L1023) but I needed to hoist this logic out because resolving layout constraints happens in `run_node`

zou3519 replied:
I ended up adding new logic to register the layout constraint once to make it clearer what is going on. Relying on the lowering being populated seemed like too much mental overload

eellison commented on torch/_inductor/graph.py:
```diff
@@ -1244,27 +1268,37 @@ def debug(msg: str) -> None:
             n
         ):
+            if is_call_function:
+                self.maybe_add_custom_op_layout_constraints(n.target)  # type: ignore[arg-type]
```
nit: could just return in this function to avoid the control flow. it already has `maybe` in its function name. it errors on some inputs but doesnt do anything on others..

zou3519 replied:
sounds good

eellison reviewed (approved):
looks great !

zou3519 commented on test/inductor/test_torchinductor_codegen_dynamic_shapes.py:
```diff
@@ -124,6 +124,7 @@ def run(*ex, **kwargs):
     "test_conv2d_channels_last_dynamic_shapes": TestFailure(("cpu",)),
     "test_conv3d_dynamic_shapes": TestFailure(("cpu",)),
     "test_conv3d_channels_last_dynamic_shapes": TestFailure(("cpu",)),
+    "test_mutable_custom_op_fixed_layout2_dynamic_shapes": TestFailure(("cpu",)),
```
Yes, I think it's because inductor doesn't generate any for loops for this code on CPU?

zou3519 committed:
Update on "[inductor] Fix needs_fixed_stride_order silent incorrectness"


Fixes #128084

The approach is option 2 of what Elias suggested in the comment
thread:
- We require tensors to have the correct stride at usage. This may
  involve a clone; if there was a clone and then a mutation into it
  then we copy_ back the result of the mutation.

The reason why I went this approach was because it was the easiest and
Inductor already works really hard to remove additional clones/copy_.

There are some cases that this doesn't generate efficient code for; for
example, if the tensor is a view, we don't change the base of the view
to have the right stride order, instead we do a clone.
The view case isn't very common so I'm ignoring it for now but we could
improve this in the future.

Test Plan:
- tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] Fix needs_fixed_stride_order silent incorrectness"


Fixes #128084

The approach is option 2 of what Elias suggested in the comment
thread:
- We require tensors to have the correct stride at usage. This may
  involve a clone; if there was a clone and then a mutation into it
  then we copy_ back the result of the mutation.

The reason why I went this approach was because it was the easiest and
Inductor already works really hard to remove additional clones/copy_.

There are some cases that this doesn't generate efficient code for; for
example, if the tensor is a view, we don't change the base of the view
to have the right stride order, instead we do a clone.
The view case isn't very common so I'm ignoring it for now but we could
improve this in the future.

Test Plan:
- tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "rocm infinite queue"

jeanschmidt commented:
@pytorchbot revert -m "Broke internal fbgemm signals, see [D61670495](https://www.internalfb.com/diff/D61670495)" -c ghfirst

zou3519 commented:
@pytorchbot merge -f "this just needs an internal-only change (D61722869), going to force merge this to avoid needing to go through another development cycle"
